tt_metal/host_api.hpp

```diff
@@ -652,6 +652,68 @@ void EnqueueReadBuffer(
+ * | region         | The region of the buffer that we are reading from                                 | const BufferRegion                  |                                    | Yes      |
+ * | blocking       | Whether or not this is a blocking operation                                       | bool                                |                                    | Yes      |
+ * | sub_device_ids | The sub-device ids to wait for completion on. If empty, waits for all sub-devices | tt::stl::Span<const uint32_t>       |                                    | No       |
+ */
+// clang-format on
+void EnqueueReadSubBuffer(
+    CommandQueue& cq,
+    std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer,
+    void* dst,
+    const BufferRegion region,
+    bool blocking,
+    tt::stl::Span<const SubDeviceId> sub_device_ids = {});
+
+// clang-format off
+/**
+ * Reads a specified region of the buffer from the device
+ *
+ * Return value: void
+ *
+ * | Argument       | Description                                                                       | Type                                | Valid Range                        | Required |
+ * |----------------|-----------------------------------------------------------------------------------|-------------------------------------|------------------------------------|----------|
+ * | cq             | The command queue object which dispatches the command to the hardware             | CommandQueue &                      |                                    | Yes      |
+ * | buffer         | The device buffer we are reading from                                             | Buffer & or std::shared_ptr<Buffer> |                                    | Yes      |
+ * | dst            | The vector where the results that are read will be stored                         | vector<DType> &                     |                                    | Yes      |
+ * | region         | The region of the buffer that we are reading from                                 | const BufferRegion                  |                                    | Yes      |
+ * | blocking       | Whether or not this is a blocking operation                                       | bool                                |                                    | Yes      |
+ * | sub_device_ids | The sub-device ids to wait for completion on. If empty, waits for all sub-devices | tt::stl::Span<const uint32_t>       |                                    | No       |
+ */
+// clang-format on
+template <typename DType>
+void EnqueueReadSubBuffer(
+    CommandQueue& cq,
+    Buffer& buffer,
+    std::vector<DType>& dst,
+    const BufferRegion region,
+    bool blocking,
+    tt::stl::Span<const SubDeviceId> sub_device_ids = {}) {
+    dst.resize(region.size / sizeof(DType));
+    EnqueueReadSubBuffer(cq, buffer, static_cast<void*>(dst.data()), region, blocking, sub_device_ids);
```

Comment:
don't we need an empty check here on the std::vector? dst.data() is used without checking if dst is empty